### PR TITLE
Add acceptable options argument to verification

### PR DESCRIPTION
**Pull request by Al-Kindi-0**
Addresses #160

### CODE REVIEW

File: verifier/src/errors.rs [hunk 1]
```diff
@@ -74,6 +77,13 @@ impl fmt::Display for VerifierError {
             Self::FriVerificationFailed(err) => {
                 write!(f, "verification of low-degree proof failed: {err}")
             }
+            Self::InsufficientConjecturedSecurity(minimal_security,proofs_security)=> {
+                write!(f, "verification failed due to parameters not providing enough conjectured security: expected at least {minimal_security} bits of security but the proof provides only {proofs_security} bits of conjectured security")
+            }
```
Comment (irakliyk) [inline verifier/src/errors.rs:82]: nit: I would probable be a bit more concise than this. For example: "insufficient proof security level: expected at least {minimal_security} bits of conjectured security, but was {proof_security} bits"

Also, I would add a space after the comma in "minimal_security,proofs_security" on line 80.

File: verifier/src/errors.rs [hunk 2]
```diff
@@ -74,6 +77,13 @@ impl fmt::Display for VerifierError {
...
+            Self::InsufficientProvenSecurity(minimal_security,proofs_security)=> {
+                write!(f, "verification failed due to parameters not providing enough proven security: expected at least {minimal_security} bits of security but the proof provides only {proofs_security} bits of proven security")
```
Comment (irakliyk) [inline verifier/src/errors.rs:84]: Similar comment as above.

File: verifier/src/errors.rs [hunk 3]
```diff
@@ -74,6 +77,13 @@ impl fmt::Display for VerifierError {
...
+            }
+            Self::UnacceptableProofOptions => {write!(f, "verification failed due to parameters not providing enough security: the proof was generated with a set of parameters not accepted by the verifier")}
```
Comment (irakliyk) [inline verifier/src/errors.rs:86]: I would rephrase it a bit since this could fail not only if security is too low but also if the security level is higher or the same, but just with a different parameter set. Maybe something like: "invalid proof options: security parameters do not match the acceptable parameter set".

File: verifier/src/lib.rs [hunk 4]
```diff
@@ -287,3 +317,13 @@ where
...
 }
+
+// ACCEPTABLE OPTIONS
+// ================================================================================================
+// Specifies either the minimal, conjectured or proven, security level or a set of
+// `ProofOptions` that are acceptable by the verification procedure.
+pub enum AcceptableOptions {
+    MinConjecturedSecurity(u32),
+    MinProvenSecurity(u32),
+    OptionSet(Vec<ProofOptions>),
```
Comment (irakliyk) [inline verifier/src/lib.rs:328]: Since this is a part of the public API, we should add doc comments for each variant.

Also, to get rid of the errors in `no_std`, we should import `Vec` form `utils::collections`.

File: verifier/src/lib.rs [hunk 5]
```diff
@@ -70,25 +70,55 @@ pub use errors::VerifierError;
...
+                        proof_security,
+                    ));
+                }
+            }
+            AcceptableOptions::OptionSet(options) => {
+                if !options.iter().any(|opt| opt == proof.options()) {
+                    return Err(VerifierError::UnacceptableProofOptions);
+                }
+            }
+        }
```
Comment (irakliyk) [inline verifier/src/lib.rs:120]: I would probably move this logic into a function implemented on `AcceptableOptions`. Maybe something like:

```Rust
impl AcceptableOptions {
    pub fn validate(&self, proof: &StarkProof) -> Result<(), VerifierError> {
        ...
    }
}
```

**Review by irakliyk — CHANGES_REQUESTED** [review]
Looks good! Thank you! I left a few small comments inline.

File: verifier/src/lib.rs [hunk 6]
```diff
@@ -70,25 +71,31 @@ pub use errors::VerifierError;
...
+    acceptable_options: &AcceptableOptions,
 ) -> Result<(), VerifierError> 
 where 
     AIR: Air, 
     HashFn: ElementHasher<BaseField = AIR::BaseField>,
     RandCoin: RandomCoin<BaseField = AIR::BaseField, Hasher = HashFn>,
 {
+    // check that `proof` was generated with an acceptable set of parameters from the point of view
+    // of the verifier
+    let _ = acceptable_options.validate::<HashFn>(&proof);
```
[the remark this answers is not in the document]
Comment (Al-Kindi-0) [reply verifier/src/lib.rs]: My bad! Fixed now!

File: verifier/src/lib.rs [hunk 7]
```diff
@@ -287,3 +294,47 @@ where
...
+    /// Minimal acceptable conjectured security level
+    MinConjecturedSecurity(u32),
+    /// Minimal acceptable proven security level
+    MinProvenSecurity(u32),
+    /// Set of acceptable proof parameters
+    OptionSet(Vec<ProofOptions>),
+}
+
+impl AcceptableOptions {
+    pub fn validate<H: Hasher>(&self, proof: &StarkProof) -> Result<(), VerifierError> {
```
[the remark this answers is not in the document]
Comment (Al-Kindi-0) [reply verifier/src/lib.rs:313]: Done!

**Review by irakliyk — APPROVED** [review]
All looks good! Thank you!
